src/cargo/ops/cargo_add/mod.rs

```diff
@@ -537,19 +538,67 @@ fn get_latest_dependency(
                     std::task::Poll::Pending => registry.block_until_ready()?,
                 }
             };
-            let latest = possibilities
-                .iter()
-                .max_by_key(|s| {
-                    // Fallback to a pre-release if no official release is available by sorting them as
-                    // less.
+
+            fn find_latest<'a>(
+                possibilities: impl Iterator<Item = &'a Summary>,
+            ) -> Option<&'a Summary> {
+                possibilities.max_by_key(|s| {
+                    // Fallback to a pre-release if no official release is available by sorting them
+                    // as less.
                     let stable = s.version().pre.is_empty();
                     (stable, s.version())
                 })
-                .ok_or_else(|| {
-                    anyhow::format_err!(
-                        "the crate `{dependency}` could not be found in registry index."
-                    )
-                })?;
+            }
+
+            let mut latest = find_latest(possibilities.iter()).ok_or_else(|| {
+                anyhow::format_err!(
+                    "the crate `{dependency}` could not be found in registry index."
+                )
+            })?;
+
+            if config.cli_unstable().msrv_policy {
+                if let Some(rust_version) = manifest
+                    .data
+                    .as_table()
+                    .get("package")
+                    .and_then(|p| p.get("rust-version"))
+                    .and_then(|v| v.as_str())
+                    .and_then(|v| semver::VersionReq::parse(v).ok())
+                {
+                    // Find the latest version of the dep which has a compatible MSRV. Candidates
+                    // without a rust-version field are treated as compatible.
+                    let latest_msrv = find_latest(possibilities.iter().filter(|s| {
+                        s.rust_version()
+                            .map(|v| v.as_str())
+                            .map(|v| semver::VersionReq::parse(v).unwrap())
+                            .map(|v| {
+                                // Rust versions need at least a major and a minor value, and cannot
+                                // contain semver operators or pre-release identifiers, so the
+                                // comparison is simple.
```

Comment:
> Rust versions need at least a major and a minor value, and cannot

While it won't make a difference to this code, minor is optional I think.

I'd recommend framing this in terms of why this is being said.  For example:
```rust
/// HACK: `rust-version` is a subset of the `VersionReq` syntax but we need to compare them with each other.  Thankfully that subset guarantees there is only ever one comparator and the operator is default, making this easier
```

Even better would to be to encode this comment in the code to help catch bugs if assumptions change
```rust
assert!(rust_version.comparators.len() == 1, "`package.rust-version = \"{rust_version}\" should just be a version with optional minor, patch");
let rust_version = &rust_version.comparators[0];
assert_eq!(rust_version.op, Op::Caret, "`package.rust-version = \"{rust_version}\" should just be a version with optional minor, patch");
assert_eq!(rust_version.prerelease, Prerelease::Empty, "`package.rust-version = \"{rust_version}\" should just be a version with optional minor, patch");
```